### PR TITLE
feat: Ability to configure multiple on request handlers

### LINKED ISSUE #1271: feat: ability to configure request handler
**Current**
We can only setup one request handler in a JS worker file. This proposal intends to break that restriction

**Proposal**

```graphql
type schema @upstream(onRequest: "onRequest") {
  # ... 
}

type Query {
  foo: Foo @http(path: "/foo", onRequest: "onRequestFoo")
  bar: Bar @http(path: "/bar", onRequest: "onRequestBar")
  baz: Bar @http(path: "/baz")
}
```

```js
function onRequest (request) {
  // 
}

function onRequestFoo (request) {
  // 
}

function onRequestBar (request) {
  // 
}

```

In the above implementation when:
1. We query `foo`, `onRequestFoo` is called with the request params.
2. We query `bar`, `onRequestBar` is called with the request params.
3. We query `baz`, `onRequest` (default handler) is called with the request params.

### PR DESCRIPTION
**Pull request by webbdays**
**Summary:**  
Ability to configure multiple on request handlers.

**Issue Reference(s):**  
Fixes #1271
/claim #1271

**Build & Testing:**

- [x] I ran `cargo test` successfully.
- [x] I have run `./lint.sh --mode=fix` to fix all linting issues raised by `./lint.sh --mode=check`.

**Checklist:**

- [x] I have added relevant unit & integration tests.
- [ ] I have updated the [documentation] accordingly.
- [x] I have performed a self-review of my code.
- [x] PR follows the naming convention of `<type>(<optional scope>): <title>`
- [ ] Fix https://github.com/tailcallhq/tailcall/issues/1570

[documentation]: https://github.com/tailcallhq/tailcallhq.github.io/tree/develop/docs

## Summary by CodeRabbit

- **New Features**
	- Introduced enhanced request handling capabilities in the GraphQL server, allowing for dynamic customization based on specific request headers.
	- Added logging functionality for better visibility into HTTP method and path used in requests.
- **Bug Fixes**
	- Updated test scenarios to reflect changes in request handling and URL updates for accuracy.
- **Refactor**
	- Implemented the use of `Mustache` templates for more flexible header management.
	- Refined HTTP and GRPC request execution logic to support additional parameters and customization.
- **Documentation**
	- Documented the ability to handle multiple `onRequest` handlers for JavaScript requests, enhancing flexibility in request processing.

### CODE REVIEW

File: src/config/config.rs [hunk 1]
```diff
@@ -540,6 +540,9 @@ impl MergeRight for Union {
 /// REST API. In this scenario, the GraphQL server will make a GET request to
 /// the API endpoint specified when the `users` field is queried.
 pub struct Http {
+    #[serde(rename = "onRequest", default, skip_serializing_if = "is_default")]
+    pub on_request: Option<String>,
+
```
[the remark this answers is not in the document]
Comment (webbdays) [reply src/config/config.rs]: Comment for onrequest field.

[the remark this answers is not in the document]
Comment (webbdays) [reply src/config/config.rs]: It's done.

File: src/config/upstream.rs [hunk 2]
```diff
@@ -32,6 +32,11 @@ pub struct Proxy {
 /// upstream server connection. This includes settings like connection timeouts,
 /// keep-alive intervals, and more. If not specified, default values are used.
 pub struct Upstream {
+    #[serde(rename = "onRequest", default, skip_serializing_if = "is_default")]
+    /// onRequest field gives the ability to specify the global request
+    /// interception handler.
+    pub global_on_request: Option<String>,
```
Comment (tusharmath) [inline src/config/upstream.rs:38]:
```suggestion
    pub on_request: Option<String>,
```

File: examples/rest-api.graphql [hunk 3]
```diff
@@ -1,12 +1,13 @@
...
-  @link(type: Operation, src: "operations/routes.graphql") {
+  @upstream(onRequest: "onRequest", baseURL: "http://jsonplaceholder.typicode.com", httpCache: true)
+  @link(type: Operation, src: "operations/routes.graphql")
+  @link(type: Script, src: "scripts/echo.js") {
   query: Query
 }
 
 type Query {
-  posts: [Post] @http(path: "/posts")
+  posts: [Post] @http(path: "/posts", onRequest: "postsRequestHandler")
```
Comment (tusharmath) [inline examples/rest-api.graphql:10]: Instead of making changes here, can you add some tests, like we have in the tests folder for JS.

Comment (webbdays) [reply examples/rest-api.graphql:10]: Sure.
Please point the tests folder location.

File: src/config/upstream.rs (hunk 2 above)
Comment (webbdays) [reply src/config/upstream.rs:38]: Sure.

File: examples/rest-api.graphql (hunk 3 above)
Comment (tusharmath) [reply examples/rest-api.graphql:10]: Check this test out for example: https://github.com/tailcallhq/tailcall/blob/main/tests/execution/test-js-request-reponse.md

**Comment by webbdays** [comment]
@coderabbitai
Help me in writing tests that cover new lines.

File: src/config/upstream.rs (hunk 2 above)
Comment (webbdays) [reply src/config/upstream.rs]: @tusharmath 
made changes.

File: examples/rest-api.graphql (hunk 3 above)
Comment (webbdays) [reply examples/rest-api.graphql:10]: @tusharmath  
added a test

**Comment by webbdays** [comment]
@tusharmath 
Can u please help me in writing tests for new lines for coverage.
Thank you.

**Comment by webbdays** [comment]
@coderabbitai
What is the tests coverage now.
Thanks.

**Comment by webbdays** [comment]
I have made changes in test now it is covering new lines.
will push the changes soon.

**Comment by tusharmath** [comment]
This was quite a tricky feature to implement. Kudos to figuring out a way to achieve it!
I think piggybacking on headers to pass the js function name might not be the right way. How about we change the trait HttpIO instead:

**Current**
```rust
pub trait HttpIO: Sync + Send + 'static {
    async fn execute(
        &self,
        request: reqwest::Request,
    ) -> anyhow::Result<Response<hyper::body::Bytes>>;
}
```
**Proposed**
```rust
pub trait HttpIO: Sync + Send + 'static {
    async fn execute(
        &self,
        request: reqwest::Request,
        filter: HttpFilter
    ) -> anyhow::Result<Response<hyper::body::Bytes>>;
}
```

```rust
#[derive(Setters, Default)]
struct HttpFilter {
  onRequest: Option<String>
}
```

We use HttpFilter so that we can extend it later with `onResponse` and other hooks.

We should store HttpFilters in the `lambda::IO::Http` and at the time of execution, we should pass the filter to the `execute` method of the client. 

You could further simplify the `blueprint::http` module by just reading from upstream and from the HTTP directive and merging it.

Let me know what you think?

**Comment by webbdays** [comment]
ok.

**Comment by tusharmath** [comment]
Converting the PR to draft. Feel free to make it ready to review when you are done.

**Comment by webbdays** [comment]
@coderabbitai
What do you say about above proposal.

**Comment by webbdays** [comment]
Hi @tusharmath ,
I have implemented the approach proposed by you.
Please look at it.
Thank you.

File: tests/server_spec.rs [hunk 4]
```diff
@@ -13,7 +13,7 @@ pub mod test {
     use reqwest_middleware::{ClientBuilder, ClientWithMiddleware};
     use tailcall::cache::InMemoryCache;
     use tailcall::cli::javascript;
-    use tailcall::http::Response;
+    use tailcall::http::{self, Response};
     use tailcall::runtime::TargetRuntime;
     use tokio::io::{AsyncReadExt, AsyncWriteExt};
```
[the remark this answers is not in the document]
Comment (webbdays) [reply tests/server_spec.rs:19]: ?

**Comment by webbdays** [comment]
@coderabbitai
update the pr summary based on the latest commit.
Thank you.

**Comment by webbdays** [comment]
Hi @tusharmath,
Please review.

**Comment by tusharmath** [comment]
@webbdays I made a few changes but am unable to push. Can you give me those permissions.

**Comment by webbdays** [comment]
sure.

File: src/cli/javascript/request_filter.rs [hunk 5]
```diff
@@ -34,14 +34,25 @@ impl RequestFilter {
...
 
     #[async_recursion::async_recursion]
-    async fn on_request(&self, mut request: reqwest::Request) -> anyhow::Result<Response<Bytes>> {
+    async fn on_request(
+        &self,
+        mut request: reqwest::Request,
+        http_filter: Option<http::HttpFilter>,
+    ) -> anyhow::Result<Response<Bytes>> {
+        // by default
+        let mut on_request = "onRequest".to_string();
```
Comment (tusharmath) [inline src/cli/javascript/request_filter.rs:43]: I think we can drop default. I understand it's a breaking change, but it's less confusing this way.

File: src/blueprint/operators/http.rs [hunk 6]
```diff
@@ -59,14 +59,41 @@ pub fn compile_http(
...
+            let mut on_request = "onRequest".to_string();
+
+            let script_links = &config_module
+                .links
+                .iter()
+                .filter(|l| l.type_of == LinkType::Script)
+                .collect::<Vec<&Link>>();
+
+            // only when we have one js script file linked via @link directive
+            if script_links.len() == 1 && script_links[0].src.ends_with(".js") {
```
[the remark this answers is not in the document]
Comment (webbdays) [reply src/blueprint/operators/http.rs:72]: Previously, when dealing with headers i have added it.
not needed now.

File: src/cli/javascript/request_filter.rs [hunk 7]
```diff
@@ -52,13 +63,13 @@ impl RequestFilter {
                         request
                             .url_mut()
                             .set_path(js_response.headers["location"].as_str());
-                        self.on_request(request).await
+                        self.on_request(request, None).await
```
Comment (tusharmath) [inline src/cli/javascript/request_filter.rs:66]: If we pass `None` here and have a default filter above, then we will be stuck in an infinite loop.

File: src/blueprint/operators/http.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (webbdays) [reply src/blueprint/operators/http.rs:72]: will remove.
may be we should add the valdiation check like without linking js file one should not pass filters.

File: src/cli/javascript/request_filter.rs (hunk 7 above)
Comment (webbdays) [reply src/cli/javascript/request_filter.rs:66]: yes, infinite loop.
Should we have empty string for http filter?

**Comment by webbdays** [comment]
if we don't want/should have any defaults,
we have one thing, user should pass  onRequest at upstream when the user links js worker file.
we let the user choose handlers.
if user doesnt link js worker file, we pass command an empty "" onRequest. which leads to None. Because it cannot find that function.
In this situation if infinite loop occurs  then its mistake in the users js handlers codes.

File: src/cli/javascript/request_filter.rs [hunk 8]
```diff
@@ -34,14 +34,25 @@ impl RequestFilter {
...
+        let mut on_request = "onRequest".to_string();
+
+        if let Some(value) = http_filter {
+            on_request = value.on_request;
+        }
+
         let js_request = JsRequest::try_from(&request)?;
         let event = Event::Request(js_request);
-        let command = self.worker.call("onRequest".to_string(), event).await?;
+        let command = self.worker.call(on_request, event).await?;
```
Comment (tusharmath) [inline src/cli/javascript/request_filter.rs:51]: worker should only be called, if the filter is not none.

File: src/config/upstream.rs [hunk 9]
```diff
@@ -165,6 +170,10 @@ impl Upstream {
     pub fn get_http_2_only(&self) -> bool {
         self.http2_only.unwrap_or(false)
     }
+
+    pub fn get_on_request(&self) -> String {
+        self.on_request.clone().unwrap_or("".to_string())
+    }
```
Comment (tusharmath) [inline src/config/upstream.rs:176]: should return an option of string.

File: src/http/mod.rs [hunk 10]
```diff
@@ -20,3 +20,8 @@ pub use request_template::RequestTemplate;
 pub use response::*;
 
 pub use crate::app_context::AppContext;
+
+#[derive(Default, Clone, Debug)]
+pub struct HttpFilter {
+    pub on_request: String,
+}
```
Comment (tusharmath) [inline src/http/mod.rs:27]: Can add a bit of documentation about what it does.

File: src/config/upstream.rs (hunk 9 above)
Comment (webbdays) [reply src/config/upstream.rs:176]: ok

File: src/cli/javascript/request_filter.rs (hunk 8 above)
Comment (webbdays) [reply src/cli/javascript/request_filter.rs:51]: ok.

File: src/http/mod.rs (hunk 10 above)
Comment (webbdays) [reply src/http/mod.rs:27]: sure.

File: src/blueprint/operators/http.rs [hunk 11]
```diff
@@ -59,14 +59,41 @@ pub fn compile_http(
             .into()
         })
         .map(|req_template| {
+            // by default
+            let mut on_request = "onRequest".to_string();
```
Comment (tusharmath) [inline src/blueprint/operators/http.rs:63]: don't need a default.

File: src/cli/javascript/request_filter.rs (hunk 8 above)
Comment (webbdays) [reply src/cli/javascript/request_filter.rs]: also add if a check that on-request is not none.

**Comment by webbdays** [comment]
do we need any more changes.

**Comment by tusharmath** [comment]
@webbdays can you update the docs here — https://github.com/tailcallhq/tailcallhq.github.io

File: src/lambda/io.rs [hunk 12]
```diff
@@ -27,6 +27,9 @@ pub enum IO {
         req_template: http::RequestTemplate,
         group_by: Option<GroupBy>,
         dl_id: Option<DataLoaderId>,
+        /// Provides us the ability to pass request
+        /// interception handler via on_request field.
+        http_filter: Option<http::HttpFilter>,
```
Comment (tusharmath) [inline src/lambda/io.rs:32]: should not be optional. Keep a default value if it's not available.

File: src/lib.rs [hunk 13]
```diff
@@ -45,10 +45,18 @@ pub trait EnvIO: Send + Sync + 'static {
 
 #[async_trait::async_trait]
 pub trait HttpIO: Sync + Send + 'static {
+    async fn execute_with<'a>(
+        &'a self,
+        request: reqwest::Request,
+        http_filter: &'a Option<http::HttpFilter>,
```
Comment (tusharmath) [inline src/lib.rs:51]:
```suggestion
        http_filter: &'a http::HttpFilter,
```
Optional isn't required.

File: src/cli/javascript/request_filter.rs [hunk 14]
```diff
@@ -65,10 +75,11 @@ impl RequestFilter {
...
 impl HttpIO for RequestFilter {
-    async fn execute(
-        &self,
+    async fn execute_with<'a>(
+        &'a self,
         request: reqwest::Request,
+        http_filter: &'a Option<http::HttpFilter>,
     ) -> anyhow::Result<Response<hyper::body::Bytes>> {
-        self.on_request(request).await
+        self.on_request(request, http_filter).await
```
Comment (tusharmath) [inline src/cli/javascript/request_filter.rs:83]: Need to check if filter.on_request is defined. If not, we should log a warning and simly call client.execute.

Comment (webbdays) [reply src/cli/javascript/request_filter.rs:83]: That i am handling in on_request.
should we need here?

**Comment by webbdays** [comment]
> @webbdays can you update the docs here — https://github.com/tailcallhq/tailcallhq.github.io

sure

File: src/cli/javascript/request_filter.rs (hunk 14 above)
Comment (webbdays) [reply src/cli/javascript/request_filter.rs:83]: let mut command = None;
        if let Some(value) = http_filter {
            if let Some(ref value) = value.on_request {
                command = self.worker.call(value.clone(), event).await?;
            }
        }

Comment (webbdays) [reply src/cli/javascript/request_filter.rs:83]: ideally need to check before we call on_request here itself. will change.

File: src/lambda/io.rs (hunk 12 above)
Comment (webbdays) [reply src/lambda/io.rs:32]: ok

File: src/cli/javascript/request_filter.rs (hunk 14 above)
Comment (webbdays) [reply src/cli/javascript/request_filter.rs:83]: should we pass on_request to on_request fn

Comment (webbdays) [reply src/cli/javascript/request_filter.rs:83]: @coderabbitai

Comment (webbdays) [reply src/cli/javascript/request_filter.rs:83]: async fn execute_with<'a>(
        &'a self,
        request: reqwest::Request,
        http_filter: &'a http::HttpFilter,
    ) -> anyhow::Result<Response<hyper::body::Bytes>> {
        if Some(http_filter.on_request) != None {
            self.on_request(request, http_filter.clone()).await
        }
        else{
            self.client.execute(request).await
        }
    }

Comment (webbdays) [reply src/cli/javascript/request_filter.rs:83]: http_filter if default
http::HttpFilter::Default()
then on_reques is None

Comment (webbdays) [reply src/cli/javascript/request_filter.rs:83]: @coderabbitai

Comment (webbdays) [reply src/cli/javascript/request_filter.rs:83]: @tusharmath 
please the correct structure of
#[derive(Default, Clone, Debug)]
/// User can configure the filter/interceptor
/// for the http requests.
pub struct HttpFilter {
    pub on_request: Option<String>,
}

File: tests/server_spec.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (webbdays) [reply tests/server_spec.rs:19]: @tusharmath

**Comment by tusharmath** [comment]
@webbdays The PR looks great! Can you please update the documentation here https://github.com/tailcallhq/tailcallhq.github.io

**Comment by tusharmath** [comment]
@webbdays converting this PR to draft until https://github.com/tailcallhq/tailcall/issues/1570 is fixed.

**Comment by webbdays** [comment]
Work still in progress.

**Comment by webbdays** [comment]
Depends on another issue.

**Comment by webbdays** [comment]
.

**Comment by tusharmath** [comment]
Ping! Would love to get this merged.

**Comment by tusharmath** [comment]
@webbdays I think this is an important PR to implement.

**Comment by webbdays** [comment]
But
#1570 is still pending.

**Comment by webbdays** [comment]
pending.

**Comment by webbdays** [comment]
pending.

**Comment by webbdays** [comment]
Hi

**Pull request closed without merging**